lib/link_preview/parser.rb

```diff
@@ -84,38 +87,74 @@ def parse_html(data)
-          image_secure_url: opengraph_image_array_first_elem['og:image:secure_url'],
-          image_url: opengraph_image_array_first_elem['og:image'] || opengraph_image_array_first_elem['og:image:url'],
-          tag: find_meta_property(doc, 'og:tag'),
-          url: find_meta_property(doc, 'og:url'),
-          type: find_meta_property(doc, 'og:type'),
-          site_name: find_meta_property(doc, 'og:site_name'),
-          video_secure_url: opengraph_video_array_first_elem['og:video:secure_url'],
-          video_url: opengraph_video_array_first_elem['og:video'] || opengraph_video_array_first_elem['og:video:url'],
-          video_type: opengraph_video_array_first_elem['og:video:type'],
-          video_width: opengraph_video_array_first_elem['og:video:width'],
-          video_height: opengraph_video_array_first_elem['og:video:height']
-        },
-        html: {
-          title: find_title(doc),
-          description: find_meta_description(doc),
-          tags: Array.wrap(find_rel_tags(doc))
-        }
+        title: find_title(doc),
+        description: find_meta_description(doc),
+        tags: Array.wrap(find_rel_tags(doc))
       }
     end
 
+    # FIXME: currently secure_url is favored over url via implicit ordering of keys
+    def parse_opengraph_common_data(doc)
+      opengraph_image_array_first_elem = find_meta_property_array(doc, 'og:image').first
+      {
+        title: find_meta_property(doc, 'og:title'),
+        description: find_meta_property(doc, 'og:description'),
+        image_secure_url: opengraph_image_array_first_elem['og:image:secure_url'],
+        image_url: opengraph_image_array_first_elem['og:image'] || opengraph_image_array_first_elem['og:image:url'],
+        tag: find_meta_property(doc, 'og:tag'),
+        url: find_meta_property(doc, 'og:url'),
+        type: find_meta_property(doc, 'og:type'),
+        site_name: find_meta_property(doc, 'og:site_name')
+      }
+    end
+
+    def parse_opengraph_video_data(doc)
+      opengraph_video_array_first_elem = find_meta_property_array(doc, 'og:video').detect { |x| x['og:video:type'] != 'text/html' }
+      parse_opengraph_common_data(doc).merge(
+        video_secure_url: opengraph_video_array_first_elem['og:video:secure_url'],
+        video_url: opengraph_video_array_first_elem['og:video'] || opengraph_video_array_first_elem['og:video:url'],
+        video_type: opengraph_video_array_first_elem['og:video:type'],
+        video_width: opengraph_video_array_first_elem['og:video:width'],
+        video_height: opengraph_video_array_first_elem['og:video:height']
+      )
+    end
+
+    def parse_opengraph_embed_data(doc)
+      return {} if ignore_opengraph_video_type_html?
+      opengraph_video_array_first_elem = find_meta_property_array(doc, 'og:video').detect { |x| x['og:video:type'] == 'text/html' }
+      opengraph_common_data = parse_opengraph_common_data(doc)
+      return {} unless opengraph_video_array_first_elem
```

Comment:
You might as well call this return before the line above since it'll be unnecessary work if the condition isn't met